### ISSUE TITLE
Birthday dates should not be subject to timezones

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#2867]
I just saw my own birthday reminder on my Friendica instance, but it shows a day before the actual day. The only explanation is that I'm on the EST timezone (GMT -04:00) and that the date that I submitted in my profile was thought as GMT.

Birth dates are a rare instance that shouldn't be subjected to timezones. I'm investigating #2865 so I can't delve into it, but I encourage anyone having the faintest idea how to exclude it from the timezone tyranny to take a stab at it.

**Comment by zotlabs** [comment]
The design of Friendica converts the datetime of the birthday person (in their own timezone) to the datetime of the observer, so that the observer says happy birthday at the correct time on the correct day. As opposed to Facebook where those of us in Australia are always wished happy birthday a day late and we wish our friends happy birthday a day early. 

I was a hero to my Australian and Asian friends on Facebook when I described my solution to them. This was filed as a Facebook bug in 2005 and still exists today. 

This feature was working correctly. If it is no longer working correctly, the bug isn't that birthdays have timezones, but that event timezones are not being converted correctly.  Your birthday should've been converted to GMT, and that is how it is stored and transmitted. Observers should see it converted from GMT to their own timezone. If you are the observer and it isn't displaying correctly, that indicates a rather serious bug in the events system; as it should be converted back to your timezone. If this is wrong, it implies that most all timezone-dependent events on the system may also be wrong since they all go through the same basic conversions to and from GMT.

Birthday reminders are an important feature in a global social network. Having them appear on the correct day is an important Friendica feature which distinguishes it from the tyranny of American companies like Facebook that refuse to acknowledge that timezones exist and that there are people on the planet who do not live in the America/Los_Angeles timezone.

**Comment by Hypolite** [comment]
Thanks for the elaboration, however I'm not sure if I understand clearly.

Let's say I was born on June 10th. This date will be stored in Friendica as an event which states that "Hypolite's birthday is from June 10th 00:00 GMT included to June 11th 00:00 GMT excluded".

My Friendica instance timezone is EST (GMT -04:00), which means that I will be shown that it is my birthday on my instance is displayed on June 9th and will be considered as today from June 9th 20:00 EST included to June 10th 20:00 EST excluded. Both displays are unexpected to me.

What I expect: my birthday date is shown as June 10th whatever the instance's timezone and is considered as today from June 10th 00:00 EST (June 10th 04:00 GMT) included to June 11th 00:00 EST (June 11th 04:00 GMT) excluded no matter the timezone as well.

**Comment by Hypolite** [comment]
I think I found the cornerstone issue. The Birthday events are created out of the `contact` table (in [include/datetime.php:555](https://github.com/friendica/friendica/blob/master/include/datetime.php#L555)), but this table doesn't hold any info on the contact's timezone. The event start and end dates are then defaulted to the UTC timezone, which prevents the expected behavior you and I describe from happening. The only people viewing birthday greetings correctly are people that set their timezone as UTC.

As a side note, I'm surprised there isn't a site-wide timezone setting. I expected the user accounts I created as Admin to have the instance's timezone, but it isn't editable anywhere, and the accounts' timezone is UTC by default.

**Comment by zotlabs** [comment]
You enter your birthday as June 10th (midnight June 9-to-10). Your timezone is set to US EST (let's say America/New_Jersey). We'll ignore the year, but let's just use 2016  So 

2016-06-10 00:00:00 is what is entered. 

It will be *stored* as GMT (June 10th, 4AM). 

2016-06-10 04:00:00

You should never see this. That's merely how it is stored. 

If you display this (your timezone is US-EST), it should be converted back to your timezone and show midnight June 9-to-10 just as you entered it. 

2016-06-10 00:00:00

If I receive this birthday notification it will display for me (Australia/Sydney) as 

2016-06-10 14:00:00 (at 2PM my time you should be at midnight). 

> I'm surprised there isn't a site-wide timezone setting.

Many sites have traditionally hosted people from across the planet. 

The stored value in the contact DB should be GMT of the "next birthday" iirc (a lot of this was changed in hubzilla however it was all working at one time in Friendica). It is transmitted to your friends from a tagged field in the feed. So 2017-06-10 04:00:00 is what I think should be in the contact table. If you view the resulting birthday event in your browser it should show 2017-06-10 00:00:00 assuming of course you are using your correct timezone.

**Comment by Hypolite** [comment]
Agreed on principles, however this is not how it is currently stored in the Friendica database.

In the `profile` table, the birthday is stored as `0000-06-10' for ageless format, and `1950-06-10` for ageful format. Which is fine since we also have the timezone available in the same table.

However, in the `contact` table where the birthday event are created from, the birthday is stored as `2017-06-10`, the next birthday date. No timezone information anymore, it becomes then implicit UTC everywhere birthday are handled.

**Comment by zotlabs** [comment]
Doing a memory refresher here... contact['bd'] is actually just for Diaspora. We used to create Friendica birthday events on the fly when parsing the feed if we noticed it was "imminent" (within the next week or so). But this code is gone. There's no longer anything to discover the contact birthday (in GMT datetime) from the feed and nothing to parse it. I have no idea where these things are being created now or if they are being created now. 

The contact['bd'] doesn't have a timezone because Diaspora doesn't have timezones.  So for them it was always UTC. It was the best we could do for that network. You may be right. So much of the original Friendica birthday timezone code and indeed the whole concept it was based on appears to be have been lost; that it may not be easy to recover at this point.

**Comment by tobiasd** [comment]
On my `/events` tab I see the birthday of @Hypolite between 6a.m. CEST the day till 6e.m. the next day. CEST is GMT+2 and EST is GMT-4 so I think those 6hour between CEST and EST are correct.

So when I look at the events at 1a.m. in the morning (CEST) I see @Hypolite has birthday in 5 hours and wont write him wishes now, because it would be too early.

**Comment by zotlabs** [comment]
So it still works - great. But there's obviously an issue somewhere w/r/t what @Hypolite sees, and I can't really help because it seems there's a completely different mechanism being used to propagate this information to your connections than what we started with. I'll have to defer this to current developers. I've looked and haven't yet been able to discover how it's currently being transmitted.

**Comment by rabuzarus** [comment]
> We used to create Friendica birthday events on the fly when parsing the feed if we noticed it was "imminent" (within the next week or so). But this code is gone

I'm not really familiar with the protocol stuff but after some research it seems that the birthday processing is done in `include/dfrn.php` (`function fetchauthor()` calls `function birthday_event()` to store the birthday in the event table)

**Comment by Hypolite** [comment]
@tobiasd This is the expected behavior. Could it be that I set my birthday before changing my timezone, which would have created the event on my instance with the UTC timezone but would have sent the correct timezone to @tobiasd's server once we connected?

**Comment by Hypolite** [comment]
Just tried to change my birthday and back again, it didn't replace the birthday event in neither of those changes.

**Comment by annando** [comment]
I think we need both dates. The one that is based on UTC to have the greetings at the right time and one that contains the exact date, since persons never say "my birthday is the 4th of July MST" but they omit the timezone. (And they will celebrate their birthday on this "4th of July" - no matter in which timezone they are currently living)

If there is currently a bug in the protocol implementation of DFRN then of course we have to fix it.

**Comment by Hypolite** [comment]
Is there any way to see what data I would share with another server?

I dug a little more, and it turns out I was born the same day as one of my Diaspora contacts. Their birthdate is stored without timezone information in the `contact`, and the created event has the same GMT-based hours as mine.

Not sure where to go from here, especially if it appears correctly on some servers but not mine.

**Comment by annando** [comment]
You could write a small script that calls ````dfrn::feed```` with the appropriate values. Then you see what is transmitted.
